autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/UpdateFailedHandler.java

```diff
@@ -51,15 +59,35 @@ public void onApplicationEvent(UpdateFailedEvent event) {
         if (failed < RETRY_THRESHOLD) {
             updateFailures.put(autoscaleClusterId, failed);
             LOGGER.debug("Increased failed count '{}' for cluster '{}'", failed, cluster.getStackCrn());
+            // TODO: Determine what needs to be done if ForbiddenException occurs with endUser.
+            if (event.getCausedBy() instanceof ForbiddenException) {
+                Integer forbiddencount = getForbiddenFailureCount(autoscaleClusterId) + 1;
+                UpdateFailedDetails updateFailedDetails = new UpdateFailedDetails(event.getLastExceptionTimestamp(), forbiddencount.longValue(),
+                        event.isWithMachineUser());
+                clusterService.setUpdateFailedDetails(cluster.getId(), updateFailedDetails);
+                forbiddenFailures.put(autoscaleClusterId, forbiddencount);
+                LOGGER.debug("Increased forbidden count '{}' for cluster '{}' with failure details: {}", forbiddencount, cluster.getStackCrn(),
+                        updateFailedDetails);
+            }
         } else {
+            if (getForbiddenFailureCount(autoscaleClusterId) >= (RETRY_THRESHOLD - 1)) {
```

Comment:
There's a corner case, where if ForbiddenFailures + non ForbiddenFailures add up to the RETRY_THRESHOLD - this code block never executes.

Instead - a simple change here would be to say, if there's any forbidden failures / the latest failure is a forbidden failure when the threshold is hit - re-initialize the machine user.

Reply:
Updated this to re-initialise if there's any forbiddenFailure that has occured.